kraft-jennifer-ndvi-automation.py

```diff
@@ -0,0 +1,698 @@
+
+# In[4]:
+
+
+### DO NOT REMOVE THIS LINE OR EDIT / MOVE THIS CELL ###
+start_time = datetime.now()
+
+
+# In[5]:
+
+
+# In this cell place all of the functions needed to run your notebook
+# You will be graded here on function application, docstrings, efficiency so ensure
+# All functions are placed here!
+
+# YOUR CODE HERE
+
+# Define function for opening and cleaning
+def open_clean_bands(band_path,
+                     crop_extent,
+                     valid_range=(0, 10000)):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    band_path : string
+        A path to the array to be opened
+    crop-extent : geopandas.dataframe
+        shape file 2d array used to clip tif arrays
+    valid_range : tuple (optional)
+        A tuple of min and max range of values for the data. Default = None
+
+
+    Returns
+    -----------
+    arr : xarray DataArray
+        An xarray DataArray with values that should be masked set to 1 for True (Boolean)
+    """
+    # TODO add tests to ensure the arrays are the same .shape
+    band = rxr.open_rasterio(band_path, masked=True).rio.clip(crop_extent.geometry,
+                                                              from_disk=True).squeeze()
+
+    # Only run this step if a valid range tuple is provided
+    if valid_range:
+        mask = ((band < valid_range[0]) | (band > valid_range[1]))
+        band = band.where(~xr.where(mask, True, False))
+
+    return band
+
+#Define function for cloud masking
+def mask_crop_ndvi(all_bands,
+                   crop_bound,
+                   pixel_qa,
+                   vals):
+    """Open and mask a single landsat band using a pixel_qa layer.
+
+    Parameters
+    -----------
+    all_bands : list
+        a list containing the xarray objects for landsat  bands 4 and  5
+    crop_bound: geopandas GeoDataFrame
+        A geopandas dataframe to be used to crop the raster data using rasterio mask().
+    pixel_qa: xarray DataArray
+        An xarray DataArray with pixel qa values that have not yet been turned into a mask (0s and 1s)
+    vals: list
+        A list of values needed to create the cloud mask
+
+
+    Returns
+    -----------
+    ndvi_crop : Xarray Dataset
+        a cropped and masked xarray object containing NDVI values
+    """
+
+    crop_json = crop_bound.geometry
+
+    # Clip pixel qa cloud mask layer
+    cl_mask_crop = pixel_qa.rio.clip(crop_json)
+
+    # Calculate NDVI
+    ndvi_xr = (all_bands[1]-all_bands[0]) / (all_bands[1]+all_bands[0])
+    # Clip NDVI layer
+    ndvi_crop = ndvi_xr.rio.clip(crop_json)
+    
+    # Apply cloud mask to NDVI
+    ndvi_crop = ndvi_crop.where(~cl_mask_crop.isin(vals))
+
+    return ndvi_crop
+
+
+# In[6]:
+
+
+# Get a list of each directory
+path = os.path.join("ndvi-automation", "sites")
+
+# Get a list of both site directories (We will talk more about automation next week)
+sites = glob(path + "/*/")
+sites
```

Comment:
You call you're variables pretty frequently after you create them. I understand this is necessary while working through the code to make sure you're variables work, but I don't think they need to be there in the final code.

Reply:
You're right! I think I tend to leave them in because I have never coded at all before this class, and have been thinking I should "show my work" and thought process/self checks throughout the assignment. But this would be a bad habit to retain outside of an educational environment, and probably really is not needed even here. Thanks for pointing this out to me! :)